test/stdcpp/src/new_test.d

```diff
@@ -32,3 +32,64 @@ unittest
     s.c = cast(MyStruct*)__cpp_new(MyStruct.sizeof);
     cpp_delete(s);
 }
+
+extern(C++) class MyClassNoGC
+{
+    __gshared int numDeleted;
+    int x = 3;
+    this(int x) @nogc { this.x = x; }
+    ~this() @nogc { ++numDeleted; }
+}
+
+@nogc unittest
+{
+    // TEST NOT COMPILING: compiler can't tell dtor of MyClassNoGC is @nogc.
+    if (0)
+        MyClassNoGC.init.__xdtor();
+}
```

Comment:
The problem wasn't in `destroy` it's here:

```console
test\stdcpp\src\new_test.d(48): Error: @nogc function new_test.__unittest_L44_C7 cannot call non-@nogc destructor new_test.MyClassNoGC.~this
```

Reply:
`__xdtor` is not the destructor; that's actually a function synthesised by the compiler.
Maybe there's a compiler bug where that function should be inferring some attributes but it's not?
It will also execute all the field destructors, and the super destructor, so if any of those are not `@nogc`, then that may be breaking the inference...?